pkg/cloudprovider/mesos/mesos.go

```diff
@@ -0,0 +1,238 @@
+		config.MesosHttpClientTimeout.Duration,
+		config.StateCacheTTL.Duration); err != nil {
+		log.V(1).Infof("failed to create mesos cloud client: %v", err)
+		return nil, err
+	} else {
+		return &MesosCloud{client: cl, config: config}, nil
+	}
+}
+
+// Instances returns a copy of the Mesos cloud Instances implementation.
+// Mesos natively provides minimal cloud-type resources. More robust cloud
+// support requires a combination of Mesos and cloud-specific knowledge.
+func (c *MesosCloud) Instances() (cloudprovider.Instances, bool) {
+	return c, true
+}
+
+// TCPLoadBalancer always returns nil, false in this implementation.
+// Mesos does not provide any type of native load balancing by default,
+// so this implementation always returns (nil, false).
+func (c *MesosCloud) TCPLoadBalancer() (cloudprovider.TCPLoadBalancer, bool) {
+	return nil, false
+}
+
+// Zones always returns nil, false in this implementation.
+// Mesos does not provide any type of native region or zone awareness,
+// so this implementation always returns (nil, false).
+func (c *MesosCloud) Zones() (cloudprovider.Zones, bool) {
+	return nil, false
+}
+
+// Clusters returns a copy of the Mesos cloud Clusters implementation.
+// Mesos does not provide support for multiple clusters.
+func (c *MesosCloud) Clusters() (cloudprovider.Clusters, bool) {
+	return c, true
+}
+
+// ListClusters lists the names of the available Mesos clusters.
+func (c *MesosCloud) ListClusters() ([]string, error) {
+	// Always returns a single cluster (this one!)
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	name, err := c.client.clusterName(ctx)
+	return []string{name}, err
+}
+
+// Master gets back the address (either DNS name or IP address) of the leading Mesos master node for the cluster.
+func (c *MesosCloud) Master(clusterName string) (string, error) {
+	clusters, err := c.ListClusters()
+	if err != nil {
+		return "", err
+	}
+	for _, name := range clusters {
+		if name == clusterName {
+			if c.client.master == "" {
+				return "", errors.New("The currently leading master is unknown.")
+			}
+
+			host, _, err := net.SplitHostPort(c.client.master)
+			if err != nil {
+				return "", err
+			}
+
+			return host, nil
+		}
+	}
+	return "", errors.New(fmt.Sprintf("The supplied cluster '%v' does not exist", clusterName))
+}
+
+// ipAddress returns an IP address of the specified instance.
+func ipAddress(name string) (net.IP, error) {
+	if name == "" {
+		return nil, noHostNameSpecified
+	}
+	ipaddr := net.ParseIP(name)
+	if ipaddr != nil {
+		return ipaddr, nil
+	}
+	iplist, err := net.LookupIP(name)
+	if err != nil {
+		log.V(2).Infof("failed to resolve IP from host name '%v': %v", name, err)
+		return nil, err
+	}
+	ipaddr = iplist[0]
+	log.V(2).Infof("resolved host '%v' to '%v'", name, ipaddr)
+	return ipaddr, nil
+}
+
+// ExternalID returns the cloud provider ID of the specified instance.
+func (c *MesosCloud) ExternalID(instance string) (string, error) {
```

Comment:
actually *MesosCloud needs to be the receiver to implement
cloudprovider.Instances


On Thu, May 14, 2015 at 5:10 AM, David Oppenheimer <notifications@github.com
> wrote:

> In pkg/cloudprovider/mesos/mesos.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8053#discussion_r30307344>
> :
>
> > +	ipaddr := net.ParseIP(name)
> > +	if ipaddr != nil {
> > +		return ipaddr, nil
> > +	}
> > +	iplist, err := net.LookupIP(name)
> > +	if err != nil {
> > +		log.V(2).Infof("failed to resolve IP from host name '%v': %v", name, err)
> > +		return nil, err
> > +	}
> > +	ipaddr = iplist[0]
> > +	log.V(2).Infof("resolved host '%v' to '%v'", name, ipaddr)
> > +	return ipaddr, nil
> > +}
> > +
> > +// ExternalID returns the cloud provider ID of the specified instance.
> > +func (c *MesosCloud) ExternalID(instance string) (string, error) {
>
> this one doesn't need to receive MesosCloud either
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8053/files#r30307344>
> .
>